src/savegame_json.cpp

```diff
@@ -1115,19 +1115,45 @@ void monster::load(JsonObject &data)
 
     data.read("hp", hp);
 
+    // sp_timeout indicates an old save, prior to the special_attacks refactor
     if (data.has_array("sp_timeout")) {
         JsonArray parray = data.get_array("sp_timeout");
         if ( !parray.empty() ) {
+            int index = 0;
             int ptimeout = 0;
-            while ( parray.has_more() ) {
+            while ( parray.has_more() && index < (int)(type->special_attacks_names.size()) ) {
                 if ( parray.read_next(ptimeout) ) {
-                    sp_timeout.push_back(ptimeout);
+                    // assume timeouts saved in same order as current monsters.json listing
+                    std::string aname = type->special_attacks_names[index++];
+                    if ( ptimeout >= 0 ) {
```

Comment:
Every time I tried to use `special_attacks[aname]`, the compiler complained about there being no 0-parameter constructor for mon_special_attack. I gave up on convincing it to do what I wanted. I'll try your solution here in the places where I'm constructing new mon_special_attack and mtype_special_attack objects and see what happens.